packages/gatsby/src/utils/page-data.js

```diff
@@ -0,0 +1,135 @@
+import { websocketManager } from "./websocket-manager"
+const fs = require(`fs-extra`)
+const path = require(`path`)
+const { store } = require(`../redux`)
+const webpackStatus = require(`./webpack-status`)
+
+const fixedPagePath = pagePath => (pagePath === `/` ? `index` : pagePath)
+
+const getFilePath = ({ publicDir }, pagePath) =>
+  path.join(publicDir, `page-data`, fixedPagePath(pagePath), `page-data.json`)
+
+const read = async ({ publicDir }, pagePath) => {
+  const filePath = getFilePath({ publicDir }, pagePath)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+  return JSON.parse(rawPageData)
+}
+
+const remove = async ({ publicDir }, pagePath) => {
+  const filePath = getFilePath({ publicDir }, pagePath)
+  if (fs.existsSync(filePath)) {
+    return await fs.remove(filePath)
+  }
+  return Promise.resolve()
+}
+
+const write = async ({ publicDir }, page) => {
+  const inputFilePath = path.join(
+    publicDir,
+    `..`,
+    `.cache`,
+    `json`,
+    `${page.path.replace(/\//g, `_`)}.json`
+  )
+  const outputFilePath = getFilePath({ publicDir }, page.path)
+  const result = await fs.readJSON(inputFilePath)
+  const body = {
+    componentChunkName: page.componentChunkName,
+    path: page.path,
+    matchPath: page.matchPath,
+    result,
+  }
+  const bodyStr = JSON.stringify(body)
+  // transform asset size to kB (from bytes) to fit 64 bit to numbers
+  const pageDataSize = Buffer.byteLength(bodyStr) / 1000
+
+  store.dispatch({
+    type: `ADD_PAGE_DATA_STATS`,
+    payload: {
+      filePath: outputFilePath,
+      size: pageDataSize,
+    },
+  })
+
+  await fs.outputFile(outputFilePath, bodyStr)
+  return body
+}
+
+let isFlushEnqueued = false
+let isFlushing = false
+
+const flush = async () => {
+  if (isFlushing) {
+    // We're already in the middle of a flush
+    return
+  }
+  isFlushing = true
```

Comment:
This should also set "isFlushEnqueued" to false.

Reply:
Done